service/storage_service.cc

```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+        // node inserts its entry, therefore we might need to wait
+        // until that happens, here.
+        return !_topology_state_machine._topology.is_empty();
+    });
+
+    while (true) {
+        std::optional<group0_guard> guard;
+        try {
+            guard.emplace(co_await _group0->client().start_operation(&_abort_source));
+        } catch (abort_requested_exception&) {
+            throw;
+        } catch (...) {
+            slogger.warn("raft topology: join_node_request: failed to start group0 operation: {}, retrying.",
+                         std::current_exception());
+        }
+        if (!guard) {
+            co_await sleep_abortable(std::chrono::seconds(1), _abort_source);
+            continue;
+        }
+
+        if (const auto *p = _topology_state_machine._topology.find(params.host_id)) {
+            const auto& rs = p->second;
+            if (rs.state == node_state::left) {
+                result.result = join_node_request_result::rejected{
+                    .reason = "The node has already been removed from the cluster",
+                };
+            } else {
+                result.result = join_node_request_result::rejected{
+                    .reason = "The node requested to join before but didn't finish the procedure. "
+                              "Please clear the data directory and restart.",
+                };
+            }
+            co_return result;
+        }
+
+        auto mutation = build_mutation_from_join_params(params, *guard);
+
+        topology_change change{{std::move(mutation)}};
+        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), *guard,
+                format("raft topology: placing join request for {}", params.host_id));
+        try {
+            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(*guard), &_abort_source);
+            break;
+        } catch (group0_concurrent_modification&) {
+            slogger.info("raft topology: join_node_request: concurrent operation is detected, retrying.");
+        }
+    }
+
+    slogger.info("raft topology: placed join request for {}", params.host_id);
+
+    // Success
+    result.result = join_node_request_result::ok {};
+    co_return result;
+}
+
+future<join_node_response_result> storage_service::join_node_response_handler(join_node_response_params params) {
+    assert(this_shard_id() == 0);
+
+    // Serialize handling the responses
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
+
+    // Wait until the join_node_request is sent.
+    co_await _join_node_request_done.get_shared_future(_abort_source);
+
+    if (_join_node_response_done.available()) {
+        // We already handled this RPC. Return immediately for idempotence.
+        slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
+        co_return join_node_response_result{};
+    }
```

Comment:
When it is possible to get more then one RPC? If a leader switches while handling the 'join` request? Worth explaining it in the comment. May be before the semaphore above.

Reply:
Added a comment at the beginning of the handler to explain why we may receive multiple RPCs (and why those can happen in parallel to each other).